### PR TITLE
Reorganise and refactor repository to prepare for integration testing.

### PR DESCRIPTION
**Pull request by evanlinjin**
## Description

This is basically a rewrite of the codebase. In summary, the codebase is now split into the following modules:

* `lnd`: Contains the lnd-client codebase.
* `scheduler`: Contains the business logic of scheduling and "payjoin-channel-funding" requests.
* `api`: Contains all the endpoint logic (currently only for HTTP).

`lnd`:
* The lnd client (now named `lnd::LndClient`) now implements `Sync` and `Send` allowing for better parallelism.
* Lnd client  "Verify funding" logic now interacts with remote lightning nodes in parallel (`lnd::LndClient::verify_funding`).
* Added `lnd::LndError` to properly catch errors instead of panicing.

`scheduler`:
* `ScheduledPayJoin` includes additional member functions for checkout output amounts and generating `open_channel` messages (from it's scheduled channels).
* `ScheduledPayJoin::test_connections` now tests connections in parallel.
* Introduce `scheduler::ProposalPsbt` (which needs a better name), that handles state for satisfying a scheduled payjoin-channel-funding.
* Introduce `SchedulerError` for handling error cases instead of panicing.
* `Scheduler` contains the business logic.

`api`:
* Handles HTTP requests with a given `Scheduler` and options.
* Introduce `ApiError` which can turn into a non-SUCCESS HTTP response instead of panicing.
* We embed static files into the binary with `include_bytes!` macro instead of requiring specification of static file location.

## For Reviewers

As mentioned by @DanGould, I should have kept this as multiple commits from the beginning to have 1 commit reflect 1 change (as this is easier on the reviewer). However, I have already merged all commits into one and splitting things out is not worth it (imo).

Instead, I have included various comments in the codebase prepended with `REPLACES:` that informs of where the associated logic is in the original code (before this PR).

## Next Steps

I think it is important to prioritize changes that will allow us to write a simple integration test with a (relatively) simple bash script. Even if we transition over to container-based integration tests, having a bash script is something to start off from.

To get there faster, I propose integrating HTTPS (we won't need to depend on a reverse proxy this way) and add associated cli flags. The fewer dependencies our bash-integration-test has, the better. This should not take long.

Then we need to add the `payjoin-client` executable into this repository.

The actual integration test should use the `btcd` and `lnd` stack (as it seems the easiest to set up).

***We need to know whether these changes break anything before merging.***

### CODE REVIEW

**Comment by evanlinjin** [comment]
@DanGould This should be two commits where:
* Commit 1: Add new files.
* Commit 2: `main.rs` changes.

Still WIP, but if you can check whether this still works in your environment, that will be great!

**Comment by DanGould** [comment]
Test fails:  ./target/debug/loptos --conf=LOPTOS.conf 10 03dde129510a7075c76cd759d1b29fea22d1ba1047a7a11b063008936931882fd0@192.168.160.7:9735 250000 10000
Error: Lnd(Generic(Status { code: Unknown, message: "already connected to peer: 03dde129510a7075c76cd759d1b29fea22d1ba1047a7a11b063008936931882fd0@192.168.160.7:48358", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None }))

File: src/main.rs [hunk 1]
```diff
@@ -1,511 +1,52 @@
...
 use crate::args::parse_args;
+use crate::scheduler::Scheduler;
 
 #[macro_use]
 extern crate serde_derive;
 extern crate configure_me;
 
 configure_me::include_config!();
 
-#[cfg(not(feature = "test_paths"))]
```
Comment (DanGould) [inline src/main.rs:25]: This was removed and not replaced. Please put that change in its own commit to document the change for review

**Comment by DanGould** [comment]
Another test error after the latest revision:

 Running `target/debug/payjoin-client 18443 none 'bitcoin:bcrt1qzw7pxt4qlvll0dg0lvxjazs6f339v9qvzsmff6?amount=0.00260420&pj=http://127.0.0.1:3000/pj
'`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Extras(PjParseError(UnsecureEndpoint))', payjoin-client/src/main.rs:30:46
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by DanGould** [comment]
Please separate this into commits each consist of 1 change and all build. That's easier for us to review and find errors

File: src/lnd.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+                tokio::spawn(async move { client.funding_state_step(req).await })
+            })
+            .collect::<Vec<_>>();
+
+        for handle in handles {
+            handle.await.unwrap()?;
+        }
+
+        Ok(())
+    }
```
Comment (evanlinjin) [inline src/lnd.rs:141]: I've made this logic actually async. We spawn a list of async tasks and collect their handlers into a vector, in which we call `await` on each element.

File: src/scheduler.rs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+            .map(|(client, node)| {
+                tokio::spawn(async move { client.ensure_node_connected(node).await })
+            })
+            .collect::<Vec<_>>();
+
+        for handle in handles {
+            handle.await.unwrap()?;
+        }
+        Ok(())
+    }
```
Comment (evanlinjin) [inline src/scheduler.rs:93]: Testing connections is now done async.

File: src/scheduler.rs [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+            };
+        }
+
+        if funding_vouts.is_empty() {
+            Err(SchedulerError::PayJoinCannotOpenAnyChannel)
+        } else {
+            let chan_ids = requests.iter().map(|(_, id, _)| *id).collect::<Vec<_>>();
+            Ok((funding_vouts, chan_ids))
+        }
+    }
```
Comment (evanlinjin) [inline src/scheduler.rs:296]: Multi-channel open is now done async.

File: src/scheduler.rs (hunk 3 above)
Comment (evanlinjin) [inline src/scheduler.rs:93]: Testing connections is now done async.

File: src/lnd.rs (hunk 2 above)
Comment (evanlinjin) [inline src/lnd.rs:141]: I've made this logic actually async. We spawn a list of async tasks and collect their handlers into a vector, in which we call `await` on each element.

File: src/scheduler.rs (hunk 3 above)
Comment (DanGould) [reply src/scheduler.rs:93]: before each connection was done serially

File: src/scheduler.rs (hunk 4 above)
Comment (DanGould) [reply src/scheduler.rs:296]: Receiver negotiates channel openings in parallel in stead of one blocking the next

File: src/lnd.rs (hunk 2 above)
Comment (DanGould) [reply src/lnd.rs:141]: Each channel funding ran before the next. This runs them in parallel

**Comment by DanGould** [comment]
Manual integration test passing ✅

The code just needs thorough review

**Comment by DanGould** [comment]
Seems like ProposalPsbt is 1 to 1 with Scheduled PayJoin. Why not just add it to Scheduled PayJoin? If it's a separate struct, I'd say ScheduledPayJoin has a Psbt rather than the other way around

**Comment by evanlinjin** [comment]
> Seems like ProposalPsbt is 1 to 1 with Scheduled PayJoin. Why not just add it to Scheduled PayJoin? If it's a separate struct, I'd say ScheduledPayJoin has a Psbt rather than the other way around, which is proposed in this pr

`ScheduledPayJoin` represents an (admin) request via `/pj/request`. It represents the channel opens that should happen when we get an original psbt (bip 78) from a sender, that contains spends to our lnd scriptPubKey (which we probably have given them in advance).

`ProposalPsbt` is badly named. It represents the "state" of operations after we receive the user's payjoin original psbt (bip 78). This includes:
1. `Scheduler::pop_payjoin` - check the original psbt to see whether an output contains the `scriptPubKey` which was provided by lnd, and if so, we know that this should trigger these channel opens (as defined in `ScheduledPayJoin`). Hence, `Scheduler::pjs` is a map of `scriptPubKey -> ScheduledPayJoin`.
2. If everything is okay, we attempt to "satisfy" the scheduled payjoin, recording state within `ProposalPsbt` (which, again, is badly named). So we do channel opens (as defined in out `ScheduledPayJoin` request) and whatnot to finalize the proposal psbt (bip 78) that we send back to the sender.

I hope that makes sense!

**Comment by DanGould** [comment]
The proposed changes have been adapted to #17
